src/devices/bus/isa/sblaster_lle.cpp

```diff
@@ -0,0 +1,381 @@
+	m_dsp->port_out_cb<3>().set(FUNC(isa8_sblaster_2_0_lle_device::dsp_port3_w));
+
+	PC_JOY(config, m_joy);
+
+	SPEAKER(config, m_speaker).front_center();
+	MC1408(config, m_dac, 0).add_route(ALL_OUTPUTS, m_speaker, 0.5);
+
+	YM3812(config, m_ym3812, ym3812_StdClock);
+	m_ym3812->add_route(ALL_OUTPUTS, m_speaker, 1.0);
+	/* no CMS */
+}
+
+static INPUT_PORTS_START( sb_dsw )
+	PORT_START("JP1")
+	PORT_CONFNAME( 0x03, 0x01, "Base Address")
+	PORT_CONFSETTING(    0x01, "220h")
+	PORT_CONFSETTING(    0x02, "240h")
+
+	PORT_START("JP4")
+	PORT_CONFNAME( 0x0f, 0x01, "IRQ Line")
+	PORT_CONFSETTING(    0x08, "IRQ2/9" )
+	PORT_CONFSETTING(    0x04, "IRQ3" )
+	PORT_CONFSETTING(    0x02, "IRQ5")
+	PORT_CONFSETTING(    0x01, "IRQ7")
+INPUT_PORTS_END
+
+ioport_constructor isa8_sblaster_2_0_lle_device::device_input_ports() const
+{
+	return INPUT_PORTS_NAME( sb_dsw );
+}
+
+isa8_sblaster_2_0_lle_device::isa8_sblaster_2_0_lle_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) :
+	device_t(mconfig, ISA8_SOUND_BLASTER_2_0_LLE, tag, owner, clock),
+	device_isa8_card_interface(mconfig, *this),
+	m_dsp(*this, "dsp"),
+	m_joy(*this, "pc_joy"),
+	m_ym3812(*this, "ym3812"),
+	m_dac(*this, "dac"),
+	m_speaker(*this, "speaker")
+{
+}
+
+void isa8_sblaster_2_0_lle_device::device_start()
+{
+	set_isa_device();
+}
+
+void isa8_sblaster_2_0_lle_device::device_reset()
+{
+	uint16_t base_port = (ioport("JP1")->read() & 0x02) ? 0x240 : 0x220;
+
+	m_isa->install_device(0x0200, 0x0207,
+		read8smo_delegate(*subdevice<pc_joy_device>("pc_joy"), FUNC(pc_joy_device::joy_port_r)),
+		write8smo_delegate(*subdevice<pc_joy_device>("pc_joy"), FUNC(pc_joy_device::joy_port_w)));
+
+	m_isa->install_device(base_port + 0x6, base_port + 0x7,
+		read8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::dsp_reset_r)),
+		write8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::dsp_reset_w)));
+	m_isa->install_device(base_port + 0x8, base_port + 0x9,
+		read8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::ym3812_16_r)),
+		write8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::ym3812_16_w)));
+	m_isa->install_device(base_port + 0xa, base_port + 0xb,
+		read8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::dsp_data_r)),
+		write8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::dsp_data_w)));
+	m_isa->install_device(base_port + 0xc, base_port + 0xd,
+		read8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::dsp_wbuf_status_r)),
+		write8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::dsp_cmd_w)));
+	m_isa->install_device(base_port + 0xe, base_port + 0xf,
+		read8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::dsp_rbuf_status_r)),
+		write8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::dsp_rbuf_status_w)));
+
+	m_isa->install_device(0x0388, 0x0389,
+		read8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::ym3812_16_r)),
+		write8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::ym3812_16_w)));
```

Comment:
You need to set a flag after installing the handlers, so that if someone flips the base address jumper and resets, you don’t end up installed at multiple addresses.

Also, can you use `emu::rw_delegate` rather than the explicit types?